lib/khovratovich/index.js

```diff
@@ -1,40 +1,162 @@
+/**
+ * Equihash for Node.js.
+ * khovratovich engine.
+ *
+ * Copyright (c) 2017 Digital Bazaar, Inc.
+ *
+ * MIT License
+ * <https://github.com/digitalbazaar/equihash/blob/master/LICENSE>
+ */
 const addon = require('bindings')('khovratovich');
 
-exports.solve = (input, options, callback) => {
-  const parameters = {
-    n: options.n || 90,
-    k: options.k || 5,
-    seed: input
-  };
+// blake2b constant
+exports.PERSONALBYTES = 16;
 
-  if(parameters.k < 1 || parameters.k > 7) {
-    // k must be less than 7
+exports.solve = (seed, options, callback) => {
+  const n = 'n' in options ? options.n : 90;
+  const k = 'k' in options ? options.k : 5;
+  const nonceLength = 'nonceLength' in options ? options.nonceLength : 4;
+  const personal =
+    'personal' in options ? options.personal :
+    Buffer.alloc(exports.PERSONALBYTES, 0);
+
+  // k must be in range of [1,7]
+  if(!(k >= 1 && k <= 7)) {
     // TODO: Find out why the implementation requires k < 7
     return callback(
-      new Error('Equihash \'k\' parameter must be between 1 and 7.'));
+      new Error('Equihash \'k\' parameter must be from 1 to 7.'));
+  }
+
+  if((n / (k + 1)) > 32) {
+    return callback(
+      new Error('Equihash \'n\' and \'k\' must satisfy n/(k+1) <= 32.'));
+  }
+
+  if(personal.length > exports.PERSONALBYTES) {
+    return callback(
+      new Error('Equihash \'personal\' length too large.'));
+  }
+
+  if(nonceLength < 4) {
+    return callback(
+      new Error('Equihash \'nonceLength\' must be >= 4 bytes.'));
+  }
+
+  // convert nonce to buffer
+  let nonce = 'nonce' in options ? options.nonce : 1;
+  if(typeof nonce === 'number') {
+    const ab = new ArrayBuffer(nonceLength);
+    const b = Buffer.from(ab);
+    b.fill(0);
+    const dv = new DataView(ab);
+    dv.setUint32(0, nonce, true);
```

Comment:
This is big-endian, yay :+1:

Reply:
The nonce is written here as little-endian.  The nonce output is just arbitrary bytes so it doesn't really matter.  This is just initializing the starting point to solve from and using little-endian as a convention mostly because zcash was doing so with its 256 bit nonce.